toolchain/semantics/semantics_context.cpp

```diff
@@ -604,13 +626,42 @@ auto SemanticsContext::CanonicalizeStructType(ParseTree::Node parse_node,
       parse_node, SemanticsTypeId::TypeType, refs_id));
   auto type_id = semantics_ir_->AddType(node_id);
   CARBON_CHECK(canonical_types_.insert({node_id, type_id}).second);
-  canonical_struct_types_nodes_.push_back(
-      std::make_unique<StructTypeNode>(canonical_id, type_id));
-  canonical_struct_types_.InsertNode(canonical_struct_types_nodes_.back().get(),
+  canonical_types_nodes_.push_back(
+      std::make_unique<TypeNode>(canonical_id, type_id));
+  canonical_struct_types_.InsertNode(canonical_types_nodes_.back().get(),
                                      insert_pos);
   return type_id;
 }
 
+auto SemanticsContext::CanonicalizeTupleType(
+    ParseTree::Node parse_node,
+    const llvm::SmallVector<SemanticsTypeId>& type_ids) -> SemanticsTypeId {
+  llvm::FoldingSetNodeID canonical_id;
+  for (const auto& type_id : type_ids) {
+    canonical_id.AddInteger(type_id.index);
+  }
+  // If a tuple with matching fields was already created, reuse it.
+  void* insert_pos;
+  auto* node =
+      canonical_tuple_types_.FindNodeOrInsertPos(canonical_id, insert_pos);
+  if (node != nullptr) {
+    return node->type_id();
+  }
+  // The tuple type doesn't already exist, so create and store it as canonical.
+  auto type_block_id = semantics_ir_->AddTypeBlock();
+  auto& type_block = semantics_ir_->GetTypeBlock(type_block_id);
+  type_block = type_ids;
```

Comment:
If possible, it'd be great to `std::move` the vector in here to avoid an unnecessary memory allocation and copy.

Reply:
If we use `std::move` here, `type_ids` will be unusable later on. As `type_ids` is passed by reference to this function, I don't think it will be safe to use `std::move` here. Please let me know if you think otherwise!

Reply:
You can handle that by changing this function to take `type_ids` by value or by rvalue reference. In either case, the caller can pass in `std::move(their_type_ids)`, and that'll avoid the copy that's happening here. The difference is:

- If you use pass by value, then callers will implicitly make a copy if they don't use `std::move`.
- If you use pass by rvalue reference, then callers will get an error if they don't use `std::move` (and will need to manually force a copy if they want one).

Given that we never want to perform implicit and unnecessary memory allocations and copies in the toolchain, I think the second option is preferable here.